File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        HashMap<Integer, Integer> hm = new HashMap<>();
        for( int num: nums){
            hm.put(num, hm.getOrDefault(num,0)+1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> pq= new PriorityQueue<>((a,b)-> a.getValue()-b.getValue());

        for(Map.Entry<Integer, Integer> entry: hm.entrySet()){
            pq.add(entry);
            if(pq.size()>k){
                pq.poll();
            }
        }

        int[] res= new int[pq.size()];
        int index= pq.size()-1;
        while(!pq.isEmpty()){
            res[index--]=pq.poll().getKey();
        }

        return res;

    }
}

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        freq={}

        for num in nums:
            freq[num]= freq.get(num,0)+1

        heap=[]

        for key,val in freq.items():
            heapq.heappush(heap,(val,key))

            if(len(heap)>k):
                heapq.heappop(heap)
        
        res=[]
        while heap:
            x,y = heapq.heappop(heap)
            res.append(y)
        
        return res
